scripts: checkpatch.pl: refinements required to reduce false positives

**Describe the bug**
There are 2 main issues covered here with how `scripts/checkpatch.pl` enforces rules [A.4](https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html#rule-a-4-c-standard-library-usage-restrictions-in-zephyr-kernel) and [A.5](https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html#rule-a-5-c-standard-library-usage-restrictions-in-zephyr-codebase) of the coding guidelines.

**Rule A.4**

Rule A.4 restricts the Zephyr kernel from being built with ISO C functions (and some POSIX functions) other than those listed.

However, `scripts/checkpatch.pl` looks at the entire Zephyr codebase when enforcing those restrictions rather than enforcing it for (only) `kernel/` and (probably) `include/zephyr/kernel*`.

With that, `scripts/checkpatch.pl` currently generates false positives in CI for valid code changes that are not in violation of Rule A.4.

Note: The rationale for Rule A.4 is to support building the Zephyr kernel with the minimal libc.

**Rule A.5**

Rule A.5 is there to restrict use of "C standard library functions and macros" in the Zephyr codebase to that of C11, mainly because e.g. Newlib and Picolibc also provide implementations of non-ISO-C library routines bundled with their versions of the C library.

However, Zephyr's POSIX API, network subsystem, as well as conformant applications and libraries do **require** being able to declare e.g. `_POSIX_C_SOURCE` or `_XOPEN_SOURCE` (possibly gated via some Kconfig option) in order to signal to the implementation that POSIX prototypes and structures should be made available (See [IEEE 1003.1-2017](https://standards.ieee.org/ieee/1003.1/7101/), section 2.2, "Application Conformance").

We generally _do_ want to support applications (libraries, etc) that are built according to standards.

Please also mention any information which could help others to understand
the problem you're facing:
 - What target platform are you using? any
 - What have you tried to diagnose or workaround this issue? a lot (see #51211)
 - Is this a regression? No

**To Reproduce**
Steps to reproduce the behavior:
1. modify `samples/hello_world` to use `gethostname()` as in #67282
2. submit PR
3. See CI produce coding guideline errors

**Expected behavior**
Applications and libraries (outside of the kernel) should be able to depend on and select POSIX options and option groups without CI generating errors.

**Impact**
A bit of a showstopper, since whenever reviewers see compliance failures, they assume that something is wrong with the code, but in many applications (particularly those in networking, fs, posix, etc), that is not the case.

**Logs and console output**
```shell
Run if [[ ! -s "compliance.xml" ]]; then
Error: See https://docs.zephyrproject.org/latest/contribute/guidelines.html#coding-style for more details.

 API_DEFINE: do not specify a non-Zephyr API for libc
File:samples/hello_world/src/main.c
Line:10
Error: Process completed with exit code 1.
```

**Environment (please complete the following information):**
 - OS: Linux
 - Toolchain (e.g Zephyr SDK, ...): Zephyr SDK v0.16.4
 - Commit SHA or Version used: 96c20d831e

**Additional context**
<img width="906" alt="Screenshot 2024-01-07 at 2 16 26 PM" src="https://github.com/zephyrproject-rtos/zephyr/assets/983494/aa0ccf8a-0a25-438a-a4e6-71e432a5b86d">

--- a/scripts/checkpatch.pl
+++ b/scripts/checkpatch.pl
@@ -600,10 +600,8 @@ our $api_defines = qr{(?x:
 	_GNU_SOURCE|
 	_ISOC11_SOURCE|
 	_ISOC99_SOURCE|
-	_POSIX_C_SOURCE|
 	_POSIX_SOURCE|
 	_SVID_SOURCE|
-	_XOPEN_SOURCE|
 	_XOPEN_SOURCE_EXTENDED
 )};
 
@@ -6550,7 +6548,7 @@ sub process {
 
 		if ($line =~ /#\s*define\s+$api_defines/) {
 			ERROR("API_DEFINE",
-			      "do not specify a non-Zephyr API for libc\n" . "$here$rawline\n");
+			      "do not specify non-standard feature test macros for embedded code\n" . "$here$rawline\n");
 		}
 
 # check for IS_ENABLED() without CONFIG_<FOO> ($rawline for comments too)

